### PR TITLE
wip: kotlin plugin copied from scala plugin.

### PR DESCRIPTION
**Pull request by hsyed**
[WIP PR for issue](https://github.com/bazelbuild/intellij/issues/148). 

Most of the work required to prevent the Kotlin settings from being wiped is done. I haven't figured out how to setup the plugin launcher so haven't been able to test it yet -- using my existing intelij and loading the plugin is too disruptive. 

@brendandouglas could you get take this from here and get it integrated ? I'm happy to continue work on the plugin after I get the launcher working properly.

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by hsyed** [comment]
I signed it!

File: kotlin/src/com/google/idea/blaze/kotlin/sync/AlwaysPresentKotlinSyncPlugin.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Set;
+
+/**
+ * Unlike most of the kotlin-specific code, will be run even if the JetBrains kotlin plugin isn't
+ * enabled.
+ */
+public class AlwaysPresentKotlinSyncPlugin extends BlazeSyncPlugin.Adapter {
+  private static final String KOTLIN_PLUGIN_ID = "org.jetbrains.kotlin";
```
Comment (hsyed) [inline kotlin/src/com/google/idea/blaze/kotlin/sync/AlwaysPresentKotlinSyncPlugin.java]: This should be correct.

File: kotlin/src/com/google/idea/blaze/kotlin/sync/BlazeKotlinSyncPlugin.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            Module workspaceModule,
+            ModifiableRootModel workspaceModifiableModel) {
+        if (!blazeProjectData.workspaceLanguageSettings.isLanguageActive(LanguageClass.KOTLIN)) {
+            return;
+        }
+        for (Library library : ProjectLibraryTable.getInstance(project).getLibraries()) {
+            // Convert the type of the SDK library to prevent the kotlin plugin from
+            // showing the missing SDK notification.
+            if (library.getName() != null && library.getName().startsWith("kotlin-stdlib")) {
+                ExistingLibraryEditor editor = new ExistingLibraryEditor(library, null);
```
Comment (hsyed) [inline kotlin/src/com/google/idea/blaze/kotlin/sync/BlazeKotlinSyncPlugin.java]: This should be the correct library type -- there isn't one for the JVM.

File: kotlin/src/com/google/idea/blaze/kotlin/sync/source/KotlinJavaLikeLanguage.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+package com.google.idea.blaze.kotlin.sync.source;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.idea.blaze.java.sync.source.JavaLikeLanguage;
+import java.util.Set;
+
+/** Provides Java-like parts of Kotlin to the Java plugin. */
+public class KotlinJavaLikeLanguage implements JavaLikeLanguage {
+  @Override
+  public Set<String> getFileExtensions() {
```
Comment (hsyed) [inline kotlin/src/com/google/idea/blaze/kotlin/sync/source/KotlinJavaLikeLanguage.java]: There are other valid extensions but the scope of this work is just the simple kotlin java.

**Comment by brendandouglas** [comment]
This won't work without corresponding changes to the Skylark aspect, to recognize kotlin targets, and extract the required information during sync.

That would also require integration tests, to prevent this code and the scala rules diverging.

How about a smaller patch which only prevents the kotlin settings getting wiped during bazel sync? BlazeDartLibrarySource + BlazeDartSyncPlugin is an example -- it prevents dart libraries being removed during sync, but doesn't do much else.

**Comment by hsyed** [comment]
@brendandouglas I'll have a look at the dart plugin approach in the short term and see how well that ends up working. I want to do this properly though if @pcj is up for it. 

I suppose we'd need to have Kotlin rules available under the bazelbuild org, right ?.  I'll have a closer look at the Scala plugin Skylark logic.

**Pull request closed without merging**